src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,355 @@
+        }
+        let v = if fees_index == 0 {
+            sorted_fee_rates[0].fee_rate
+        } else if fees_index == percentiles.len() {
+            sorted_fee_rates.last().unwrap().fee_rate
+        } else {
+            // Notation mimics https://en.wikipedia.org/wiki/Percentile#Weighted_percentile
+            let vk = sorted_fee_rates[fees_index - 1].fee_rate;
+            let vk1 = sorted_fee_rates[fees_index].fee_rate;
+            let pk = percentiles[fees_index - 1];
+            let pk1 = percentiles[fees_index];
+            vk + (target_percentile - pk) / (pk1 - pk) * (vk1 - vk)
+        };
+        values_at_target_percentiles.push(v);
+    }
+
+    FeeRateEstimate {
+        high: values_at_target_percentiles[2],
+        middle: values_at_target_percentiles[1],
+        low: values_at_target_percentiles[0],
+    }
+}
+
+/// If the weights in `working_rates` do not add up to `full_block_weight`, add a new entry **in
+/// place** that takes up the remaining space.
+fn maybe_add_minimum_fee_rate(working_rates: &mut Vec<FeeRateAndWeight>, full_block_weight: u64) {
+    let mut total_weight = 0u64;
+    for rate_and_weight in working_rates.into_iter() {
+        total_weight = match total_weight.checked_add(rate_and_weight.weight) {
+            Some(result) => result,
+            None => return,
+        };
+    }
+
+    if total_weight < full_block_weight {
+        let weight_remaining = full_block_weight - total_weight;
+        working_rates.push(FeeRateAndWeight {
+            fee_rate: MINIMUM_TX_FEE_RATE,
+            weight: weight_remaining,
+        })
+    }
+}
+
+/// Depending on the type of the transaction, calculate fee rate and total cost.
+fn fee_rate_and_weight_from_receipt(
+    metric: &dyn CostMetric,
+    tx_receipt: &StacksTransactionReceipt,
+    block_limit: &ExecutionCost,
+) -> Option<FeeRateAndWeight> {
+    let (payload, fee, tx_size) = match tx_receipt.transaction {
+        TransactionOrigin::Stacks(ref tx) => Some((&tx.payload, tx.get_tx_fee(), tx.tx_len())),
+        TransactionOrigin::Burn(_) => None,
+    }?;
+    let scalar_cost = match payload {
+        TransactionPayload::TokenTransfer(_, _, _) => {
+            // TokenTransfers *only* contribute tx_len, and just have an empty ExecutionCost.
+            metric.from_len(tx_size)
+        }
+        TransactionPayload::Coinbase(_) => {
+            // Coinbase txs are "free", so they don't factor into the fee market.
+            return None;
+        }
+        TransactionPayload::PoisonMicroblock(_, _)
+        | TransactionPayload::ContractCall(_)
+        | TransactionPayload::SmartContract(_) => {
+            // These transaction payload types all "work" the same: they have associated ExecutionCosts
+            // and contibute to the block length limit with their tx_len
+            metric.from_cost_and_len(&tx_receipt.execution_cost, &block_limit, tx_size)
+        }
+    };
+    let denominator = if scalar_cost >= 1 {
```

Comment:
Should this be `if (scalar_cost as f64) >= MINIMUM_TX_FEE_RATE` here instead, if the point is to simply clamp `denominator` to be at least `MINIMUM_TX_FEE_RATE`?

Reply:
I think intention is that it should be positive, rather than `0`. I don't think it refers to `minimum_fee_rate`. Line 341 should probably be changed to NOT be `MINIMUM_TX_FEE_RATE`, maybe that is the mistake.

We could instead use `scalar_cost > 0` to make this clear.

I copied this from code written by @kantai  so maybe he wants to weigh in?

Reply:
I rewrote using cmp::min, maybe it's clearer now.